opendevin/runtime/client/runtime.py

```diff
@@ -0,0 +1,259 @@
+        self._init_docker(sid,container_image)
+
+    def _init_docker(self,sid,container_image):
+        self.container_image = container_image
+        # (
+        #     config.sandbox_container_image
+        #     if container_image is None
+        #     else container_image
+        # )
+        self.instance_id = (
+            sid + str(uuid.uuid4()) if sid is not None else str(uuid.uuid4())
+        )
+        self.container_name = self.container_name_prefix + self.instance_id
+        try:
+            self.docker_client = docker.from_env()
+            self._init_sandbox()
+        except Exception as ex:
+            print(
+                "Launch docker client failed. Please make sure you have installed docker and started the docker daemon."
+            )
+            raise ex
+    
+    def _init_event_stream(self,event_stream: EventStream):
+        self.event_stream = event_stream
+        self.event_stream.subscribe(EventStreamSubscriber.RUNTIME, self.on_event)
+
+    def _init_websocket(self):
+        self.websocket = None
+        # TODO: need to initialization globally only once
+        # self.loop = asyncio.new_event_loop()
+        # asyncio.set_event_loop(self.loop)
+        # self.loop.run_until_complete(self._init_websocket_connect())
+    
+    async def _init_websocket_connect(self):
+        self.websocket = await websockets.connect(self.uri)
+    
+    def _init_sandbox(self):
+        try:
+            # start the container
+            mount_dir = config.workspace_mount_path
+            self.container = self.docker_client.containers.run(
+                self.container_image,
+                command='tail -f /dev/null',
+                # TODO: test it in mac and linux
+                # network_mode='host',
+                working_dir=self.sandbox_workspace_dir,
+                name=self.container_name,
+                detach=True,
+                volumes={mount_dir: {'bind': self.sandbox_workspace_dir, 'mode': 'rw'}},
+            )
+            print('Container started')
+        except Exception as e:
+            print('Failed to start container')
+            raise e
+
+    @property
+    def sandbox_workspace_dir(self):
+        return config.workspace_mount_path_in_sandbox
+
+    def close(self):
+        containers = self.docker_client.containers.list(all=True)
+        for container in containers:
+            try:
+                if container.name.startswith(self.container_name_prefix):
+                    container.remove(force=True)
+            except docker.errors.NotFound:
+                pass
+        self.docker_client.close()
+    
+    async def on_event(self, event: Event) -> None:
+        print("EventStreamRuntime: on_event triggered")
+        if isinstance(event, Action):
+            observation = await self.run_action(event)
+            print("EventStreamRuntime: observation", observation)
+            # observation._cause = event.id  # type: ignore[attr-defined]
```

Comment:
I think we will need to fix this. I'm not at the computer so I can't test, but I think that at least, the evals output.jsonl file will not get history action, obs with the correct obs related to the action, if `cause` isn't linking them anymore.

Reply:
Get it. Currently this code are only integrated into main logic. So no evals logic will be influenced now. But yeah, we need to fix it before my integrate it into main logic.